File: amic-platform/src/modules/ma/types/dealSetup.ts
// ── 딜 셋업 AI 에이전트 타입 ─────────────────────────────

import type { DealType, TransactionSide, DealStructure } from "./transaction";

export type DDWorkstream =
  | "FDD_FINANCIAL_STATEMENTS"
  | "FDD_REVENUE"
  | "FDD_WORKING_CAPITAL"
  | "FDD_DEBT_CASH"
  | "FDD_PROJECTIONS"
  | "LDD_CORPORATE"
  | "LDD_PERMITS"
  | "LDD_CONTRACTS"
  | "LDD_ASSETS"
  | "LDD_LABOR"
  | "LDD_LITIGATION"
  | "LDD_IP"
  | "LDD_INSURANCE"
  | "LDD_ENVIRONMENT"
  | "TDD_CORPORATE_TAX"
  | "TDD_VAT"
  | "TDD_TRANSFER_PRICING"
  | "TDD_WITHHOLDING"
  | "TDD_TAX_INCENTIVES"
  | "OTHER";

export type BuyerType =
  | "STRATEGIC"
  | "FINANCIAL_SPONSOR"
  | "FAMILY_OFFICE"
  | "INDIVIDUAL"
  | "OTHER";

export type BuyerTier = "TIER_1" | "TIER_2" | "TIER_3" | null;

// ── 요청 ──────────────────────────────────────────────────

export interface DealSetupRequest {
  description: string;
  lead_advisor_email: string;
}

// ── AI 미리보기 하위 구조 ─────────────────────────────────

export interface DealSetupTransaction {
  name: string;
  deal_type: DealType;
  side: TransactionSide;
  target_company_name: string;
  client_name: string;
  estimated_deal_value: number | null;
  currency: string;
  deal_structure: DealStructure | null;
  industry: string | null;
  target_close_date: string | null;
  notes: string | null;
}

export interface DealSetupDDItem {
  workstream: DDWorkstream;
  title: string;
  description: string | null;
  due_date: string | null;
}

export interface DealSetupTimelineItem {
  event_type: string;
  title: string;
  description: string | null;
  event_date: string;
}

export interface DealSetupBuyerItem {
  company_name: string;
  buyer_type: BuyerType | null;
  tier: BuyerTier;
  notes: string | null;
}

// ── AI 미리보기 응답 ──────────────────────────────────────

export interface DealSetupPreview {
  transaction: DealSetupTransaction;
  dd_checklist: DealSetupDDItem[];
  timeline: DealSetupTimelineItem[];
  buyer_candidates: DealSetupBuyerItem[];
  cost_usd: number;
  model_used: string;
}

// ── 확인 후 저장 요청 ─────────────────────────────────────

export interface DealSetupConfirm {
  transaction: DealSetupTransaction;
  dd_checklist: DealSetupDDItem[];
  timeline: DealSetupTimelineItem[];
  buyer_candidates: DealSetupBuyerItem[];
  lead_advisor_email: string;
}

// ── 저장 결과 ─────────────────────────────────────────────

export interface DealSetupResult {
  transaction_id: string;
  dd_checklist_count: number;
  timeline_count: number;
  buyer_count: number;
}

File: deal-mgmt/app/schemas/deal_setup.py
"""딜 셋업 AI 에이전트 — Pydantic 스키마."""

from __future__ import annotations

import uuid

from pydantic import BaseModel, Field

from app.models.enums import (
    BuyerTier,
    BuyerType,
    DDWorkstream,
    DealStructure,
    DealType,
    TransactionSide,
)

# ── 요청 ──────────────────────────────────────────────────


class DealSetupRequest(BaseModel):
    """자연어 딜 설명으로 AI 미리보기 요청."""

    description: str = Field(..., min_length=10, max_length=10000)
    lead_advisor_email: str = Field(
        ...,
        max_length=255,
        pattern=r"^[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}$",
    )


# ── AI 추출 미리보기 하위 구조 ─────────────────────────────


class DealSetupTransaction(BaseModel):
    """AI가 추출한 거래 기본 정보."""

    name: str
    deal_type: DealType
    side: TransactionSide
    target_company_name: str
    client_name: str
    estimated_deal_value: float | None = None
    currency: str = "KRW"
    deal_structure: DealStructure | None = None
    industry: str | None = None
    target_close_date: str | None = None
    notes: str | None = None


class DealSetupDDItem(BaseModel):
    """AI가 생성한 DD 체크리스트 항목."""

    workstream: DDWorkstream
    title: str
    description: str | None = None
    due_date: str | None = None


class DealSetupTimelineItem(BaseModel):
    """AI가 생성한 타임라인 이벤트."""

    event_type: str
    title: str
    description: str | None = None
    event_date: str


class DealSetupBuyerItem(BaseModel):
    """AI가 생성한 매수 후보."""

    company_name: str
    buyer_type: BuyerType | None = None
    tier: BuyerTier | None = None
    notes: str | None = None


# ── AI 미리보기 응답 ──────────────────────────────────────


class DealSetupPreview(BaseModel):
    """AI가 추출한 딜 구조 미리보기 — DB 저장 전 사용자 확인용."""

    transaction: DealSetupTransaction
    dd_checklist: list[DealSetupDDItem]
    timeline: list[DealSetupTimelineItem]
    buyer_candidates: list[DealSetupBuyerItem]
    cost_usd: float
    model_used: str


# ── 확인 후 저장 요청 ─────────────────────────────────────


class DealSetupConfirm(BaseModel):
    """사용자가 편집/확인 후 DB 저장 요청."""

    transaction: DealSetupTransaction
    dd_checklist: list[DealSetupDDItem]
    timeline: list[DealSetupTimelineItem]
    buyer_candidates: list[DealSetupBuyerItem]
    lead_advisor_email: str = Field(
        ...,
        max_length=255,
        pattern=r"^[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}$",
    )


# ── 저장 결과 ─────────────────────────────────────────────


class DealSetupResult(BaseModel):
    """DB 저장 결과."""

    transaction_id: uuid.UUID
    dd_checklist_count: int
    timeline_count: int
    buyer_count: int
